cpp/experiments/employee_factory/source/facility/factory.h

```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "../role.h"
+
+namespace facility {
+
+class Factory {
+ private:
+  std::vector<std::unique_ptr<role::Employee>> employees{};
+
+ public:
+  void Accept(std::unique_ptr<role::Employee>&& employee);
```

Comment:
Short answer. Yes. I want to take a unique pointer to save it. I want to pass it as an rvalue.

Maybe I should provide context.

- I wanted to accept anonymous objects. Why?
- Code in the file `main.cpp` only needs to _"feed"_ factory with employees. Why?
- Because I don't need to use the instances of `role` anywhere else. Why?
- Because the goal was straightforward: Create a factory, let employees enter, and make them work. Why?
- I want to implement the runtime polymorphism concept. I created the story with the intent to reflect runtime polymorphism. Why?
- I wanted to get practice writing the c++ code. I also wanted to try this out: #95 

The only way to make this code work is to use the rvalue reference:

```C++
thornhil_industries.Accept(std::make_unique<role::Janitor>("John", "Doe"));
```

- We give an rvalue to the `Accept` method
- We can not copy a unique pointer
- I wanted to consume value
- I need an rvalue reference to consume it

Is there possibly a more elegant way to consume the rvalue unique pointer?

Reply:
> The only way to make this code work is to use the rvalue reference:

Rval and val approaches are very similar, both should work. This code compiles:
``` c++
void consumer_rval(std::unique_ptr<int>&& thing_to_consume);
void consumer_val(std::unique_ptr<int> thing_to_consume);

int main() {
    consumer_val(std::make_unique<int>(1));
    consumer_rval(std::make_unique<int>(2));

    auto val1 = std::make_unique<int>(1);
    auto val2 = std::make_unique<int>(2);

    consumer_val(std::move(val1));
    consumer_rval(std::move(val2));
}
```

Also, i've tested your code localy with value approach, it works fine. Just the move constructor for the value parameter gets invoked.

Very similar results, however the `consumer_val` is more often seen in practice. Up to you to choose but as I said **both** approaches compile & do what you want.

Reply:
I made a decision.

My method will accept a reference to an rvalue.

That promises two things:
- I will mutate whatever you give me.
- I accept only rvalue. For example, an anonymous instance or explicit `std::move`

Results:

- Error when passing lvalue
- I expect a move-only type
- Explicit mutability on the place of usage

Reply:
A possible reference for later _(for me)_ on [sink parameters](https://dzone.com/articles/c-sink-parameter-passing).